File: Java/프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static int answer = 0;
    static boolean[] visited;
    
    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        dfs(k, dungeons, 0);
        return answer;
    }
    
    private void dfs(int k, int[][] dungeons, int count) {
        answer = Math.max(answer, count);
        
        for (int i = 0; i < dungeons.length; i++) {
            int need = dungeons[i][0];
            int damage = dungeons[i][1];
            
            if (!visited[i] && k >= need) {
                visited[i] = true;
                dfs(k - damage, dungeons, count + 1);
                visited[i] = false;
            }
        }
    }
}

File: Python3/프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    max_count = 0
    
    for order in permutations(dungeons, len(dungeons)):
        now_cost = k
        count = 0
        for need, cost in order:
            if now_cost >= need:
                count += 1
                now_cost -= cost
            else: break
        max_count = max(max_count, count)
    
    return max_count